File: backend/tracker/tracker.py
import asyncio

from fastapi import APIRouter
from pydantic import BaseModel

router = APIRouter()


class TrackerCommand(BaseModel):
    azimuth: float
    elevation: float


tracker_state = {

    "connected": True,

    "target_azimuth": 0.0,
    "target_elevation": 0.0,

    "current_azimuth": 0.0,
    "current_elevation": 0.0
}


@router.post("/tracker")
async def set_tracker_position(
    command: TrackerCommand
):

    tracker_state["target_azimuth"] = (
        command.azimuth
    )

    tracker_state["target_elevation"] = (
        command.elevation
    )

    return {
        "status": "ok",
        "azimuth": command.azimuth,
        "elevation": command.elevation
    }


@router.get("/tracker/status")
async def get_tracker_status():

    return tracker_state


async def tracker_simulator():

    while True:

        az_diff = (
            tracker_state["target_azimuth"]
            -
            tracker_state["current_azimuth"]
        )

        if az_diff > 180:
            az_diff -= 360

        if az_diff < -180:
            az_diff += 360

        if abs(az_diff) > 1:

            tracker_state["current_azimuth"] += (
                1 if az_diff > 0 else -1
            )

        else:

            tracker_state["current_azimuth"] = (
                tracker_state["target_azimuth"]
            )

        tracker_state["current_azimuth"] %= 360

        el_diff = (
            tracker_state["target_elevation"]
            -
            tracker_state["current_elevation"]
        )

        if abs(el_diff) > 1:

            tracker_state["current_elevation"] += (
                1 if el_diff > 0 else -1
            )

        else:

            tracker_state["current_elevation"] = (
                tracker_state["target_elevation"]
            )

        await asyncio.sleep(0.05)

File: frontend/js/tracker.js
function simulateControllerFeedback() {

    AppState.controller.rxAzimuth =
        AppState.tracker.currentAzimuth;

    AppState.controller.rxElevation =
        AppState.tracker.currentElevation;

    AppState.controller.lastUpdate =
        Date.now();
}

function sendToController() {

    AppState.controller.txAzimuth =
        AppState.tracker.targetAzimuth;

    AppState.controller.txElevation =
        AppState.tracker.targetElevation;

    AppState.controller.lastUpdate =
        Date.now();
}

async function sendTrackerCommand() {

    if (!isTargetReachable()) {

        if (!targetOutOfRange) {

            console.log(
                'Target outside antenna limits'
            );

            targetOutOfRange = true;
        }

            return;
        }

    targetOutOfRange = false;

    try {

        await fetch(
            'http://localhost:8000/api/v1/tracker',
            {
                method: 'POST',

                headers: {
                    'Content-Type':
                        'application/json'
                },

                body: JSON.stringify({
                    azimuth:
                        AppState.tracker.targetAzimuth,

                    elevation:
                        AppState.tracker.targetElevation
                })
            }
        );

    } catch(error) {

        console.error(error);
    }
}

async function getTrackerStatus() {
    try {

        const response = await fetch(
            'http://localhost:8000/api/v1/tracker/status'
        );

        const data = await response.json();

        AppState.tracker.currentAzimuth =
            data.current_azimuth;

        AppState.tracker.currentElevation =
            data.current_elevation;

        console.log(AppState.tracker);

        AppState.tracker.targetAzimuth =
            data.target_azimuth;

        AppState.tracker.targetElevation =
            data.target_elevation;

    } catch(error) {

        console.error(error);
    }
}

function getAngleError(a, b) {

    let diff = Math.abs(a - b);

    if (diff > 180) {
        diff = 360 - diff;
    }

    return diff;
}

function isTargetReachable() {

   const azOk =
    AppState.tracker.targetAzimuth >=
        AppState.antenna.azimuthMin &&

    AppState.tracker.targetAzimuth <=
        AppState.antenna.azimuthMax;

const elOk =
    AppState.tracker.targetElevation >=
        AppState.antenna.elevationMin &&

    AppState.tracker.targetElevation <=
        AppState.antenna.elevationMax;

    return azOk && elOk;
}

function isNearLimit() {

    const warningZone = 10;

    return (
        AppState.tracker.targetAzimuth <=
            AppState.antenna.azimuthMin + warningZone ||

        AppState.tracker.targetAzimuth >=
            AppState.antenna.azimuthMax - warningZone
);
}